Db refactor

DeliciousHair opened the pull request:
DB code refactoring for #88

The primary goal was to remove all of the calls to `serde_json::to_string()` for the data-handling, thus enabling us to do (more or less):

```rust
Order {
    row.column,
    ...
}
```

as well as clean up the SQL for easier reading. This has mostly been accomplished, with further refinements easily accomplished once the upstream issues in `sqlx` are addressed. See #314 for issues we are tracking.

thomaseizinger commented on daemon/migrations/20210903050345_create_cfd_and_order_tables.sql:
```diff
@@ -1,19 +1,21 @@
+    leverage                        integer     not null,
+    liquidation_price               text        not null,
+    creation_timestamp_seconds      integer     not null,
+    creation_timestamp_nanoseconds  integer     not null,
```
I think I'd rather drop the nanoseconds. What do you think @da-kami ?

DeliciousHair replied:
@thomaseizinger @da-kami I've left the ns in for the moment as they are now handled without issue. Can drop them, but it looks like that may lead to a bit more work upstream as the timestamps are generated by `SystemTime::new()` so if the ns are dropped, things may not line up.

Not a big deal either way, this just seemed the path of least resistance for the moment.

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -296,8 +306,8 @@ pub async fn load_cfd_by_order_id(
     let max_quantity = serde_json::from_str(row.max_quantity.as_str()).unwrap();
```
Parsing `max_quantity` via `serde` here is not going to work is it? We are only `to_string`ing it on the insert.

DeliciousHair replied:
FYI, all things that involve `serde_json::from_str()` will go away with one exception:

```rust
    // TODO:
    // still have the use of serde_json::from_str() here, which will be dealt with
    // via https://github.com/comit-network/hermes/issues/290
    Ok(Cfd {
        order,
        quantity_usd: Usd(Decimal::from_str(&row.quantity_usd).unwrap()),
        state: serde_json::from_str(row.state.as_str()).unwrap(),
    })
```

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -493,6 +513,19 @@ pub async fn load_cfds_by_oracle_event_id(
+
+fn convert_to_duration(row_secs: i64, row_nanos: i32) -> Duration {
+    Duration::new(row_secs, row_nanos)
+}
```
Don't think this function is providing much benefit?

DeliciousHair replied:
Benefit is visual only; if nanoseconds are dropped then even that benefit is gone. :-)

thomaseizinger commented on daemon/src/model.rs:
```diff
@@ -14,7 +14,8 @@ use uuid::Uuid;
 
-#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
+#[sqlx(type_name = "String")]
```
IIRC, this only does something for postgres so we can remove it.

thomaseizinger replied:
This is unaddressed !

thomaseizinger commented on daemon/src/db_sql/query_load_cfd_by_order_id.sql:
```diff
@@ -0,0 +1,70 @@
+    state.state
+
+from order
+    inner join state on state.order_id = order.order_id
```
Missing newlines on these.

thomaseizinger commented on daemon/src/db_sql/query_load_all_cfds.sql:
```diff
@@ -0,0 +1,69 @@
+with order as (
```
I'd prefer if the changes were made in two steps:

1. Moving queries to external files
2. Changing up the query to use `with`

thomaseizinger reviewed (changes requested):
Thanks!

At the moment, there are multiple changes mangled up here. I can already see that this will prove tricky to untangle into multiple, distinct commits.

I would suggest the following:

Make a new branch off master where you just change the SQL queries to use the `with` syntax and move them into separate files without any of the improvements around `sqlx::Type` etc.

Given that you already have the queries, that should be reasonably easy. On top of that, we can then introduce `sqlx::Type` etc and get rid of all the `serde_json` stuff!

thomaseizinger commented on daemon/src/db_sql/query_load_all_cfds.sql:
```diff
@@ -0,0 +1,71 @@
+
+from ord
+  inner join state on state.order_id = ord.order_id
+  
```
Missing newline!

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -66,98 +71,39 @@ pub async fn load_order_by_id(
-    let term = serde_json::from_str(row.term.as_str()).unwrap();
-    let origin = serde_json::from_str(row.origin.as_str()).unwrap();
-    let oracle_event_id = OracleEventId(row.oracle_event_id);
+    let row = sqlx::query_file!("src/db_sql/query_load_order_by_id.sql", id)
```
I'd prefer if all the SQL files are next to the rust files and not in their own directory.

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -66,98 +71,39 @@ pub async fn load_order_by_id(
+        id: row.uuid,
+        trading_pair: row.trading_pair,
+        position: row.position,
+        price: Usd(Decimal::from_str(&row.initial_price).unwrap()),
```
I know we `unwrap()`ed here before but really we shouldn't. Can you replace those with `?` please?

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -637,7 +453,7 @@ mod tests {
 
         let cfds_from_db = load_all_cfds(&mut conn).await.unwrap();
-        let cfd_from_db = cfds_from_db.first().unwrap().clone();
+        let cfd_from_db = cfds_from_db.last().unwrap().clone();
```
Why did we need to change this test?

This looks like a regression!

DeliciousHair replied:
This was changed because the the SQL had an unnecessary reverse-ordering step:
```
        where cfd_states.state in (
            select
              state
              from cfd_states
            where cfd_id = cfds.id
            order by id desc
            limit 1
        )
```

With the reversal step removed, the test item moves from `.first()` to `.last()` is all.

thomaseizinger reviewed (changes requested):
Thank you! Massive improvement over what we currently have!

- Found what I think is a regression.
- A few unaddressed comments and commits need to be squashed together.

thomaseizinger commented on daemon/src/model/cfd.rs:
```diff
@@ -67,6 +68,7 @@ impl From<Origin> for Role {
 }
 
 /// A concrete order created by a maker for a taker
+// #[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
```
This doesn't seem necessary :)

thomaseizinger commented on Cargo.lock:
```diff
@@ -4,9 +4,9 @@ version = 3
 [[package]]
 name = "ahash"
-version = "0.7.4"
+version = "0.7.5"
```
These update must have been accidental. Can you undo that please?

`git checkout master -- Cargo.lock` should do the trick.

thomaseizinger commented on daemon/src/db.rs:
```diff
@@ -66,97 +68,83 @@ pub async fn load_order_by_id(
 }
 
 pub async fn insert_cfd(cfd: Cfd, conn: &mut PoolConnection<Sqlite>) -> anyhow::Result<()> {
-    let mut tx = conn.begin().await?;
```
Cool! No more transaction for this!

thomaseizinger commented on daemon/src/model.rs:
```diff
@@ -13,7 +13,7 @@ use uuid::Uuid;
 pub mod cfd;
 
-#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq)]
+#[derive(Debug, Copy, Clone, Serialize, Deserialize, PartialEq, sqlx::Type)]
```
I think this is not actually doing anything because we are not specifying `#[sqlx(transparent)]`.

Can you remove it to avoid confusion about what this is doing?

DeliciousHair replied:
Oops! Yes, removed now.

thomaseizinger reviewed (changes requested):
Thanks!

2 more tiny changes and this is good to go in. Just amend the now only commit for that (`git commit --amend`). Also, can you reword the commit message to something more descriptive please?

DeliciousHair committed:
DB code rebasing for #88

The primary goal was to remove all of the calls to `serde_json::to_string()`
for the data-handling, thus enabling us to do (more or less):

```rust
Order {
    row.column,
    ...
}
```

as well as clean up the SQL for easier reading. This has mostly been
accomplished, with further refinements easily accomplished once the
upstream issues in `sqlx` are addressed. See #314 for issues we are
tracking.

thomaseizinger reviewed (approved):
Thanks!

Feel free to `bors r+` once CI is green.
